### PR TITLE
[update] add drivers for stm32mp1.

### PR DESCRIPTION
**Pull request by thread-liu**
## 拉取/合并请求描述：(PR description)

[
 添加外设驱动 DAC，ADC，SPI，软/硬 I2C，WWDG，TIM，LPTIM，PWM。
所有外设驱动已在 STM32MP1-DK1 开发板上验证。
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other style
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality

### CODE REVIEW

File: bsp/stm32/libraries/HAL_Drivers/config/mp1/dac_config.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+       .Instance                      = DAC1,                          \
+    }
+#endif /* DAC1_CONFIG */
+#endif /* BSP_USING_DAC1 */
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __DAC_CONFIG_H__ */
```
Comment (SummerLife) [inline bsp/stm32/libraries/HAL_Drivers/config/mp1/dac_config.h:33]: need a new line

File: bsp/stm32/libraries/HAL_Drivers/drv_dac.c [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+//#define DRV_DEBUG
+#define LOG_TAG             "drv.dac"
+#include <drv_log.h>
+
+static DAC_HandleTypeDef dac_config[] =
+{
+#ifdef BSP_USING_DAC1
+    DAC1_CONFIG,
+#endif
+
```
Comment (SummerLife) [inline bsp/stm32/libraries/HAL_Drivers/drv_dac.c:25]: del useless line

File: bsp/stm32/stm32mp157a-st-discovery/.config [hunk 3]
```diff
@@ -108,7 +108,34 @@ CONFIG_FINSH_ARG_MAX=10
 #
 # Device virtual file system
 #
-# CONFIG_RT_USING_DFS is not set
+CONFIG_RT_USING_DFS=y
```
Comment (SummerLife) [inline bsp/stm32/stm32mp157a-st-discovery/.config:111]: disable the dfs option

File: bsp/stm32/stm32mp157a-st-discovery/rtconfig.h [hunk 4]
```diff
@@ -76,6 +76,24 @@
 
 /* Device virtual file system */
 
+#define RT_USING_DFS
```
Comment (SummerLife) [inline bsp/stm32/stm32mp157a-st-discovery/rtconfig.h:79]: ~

File: bsp/stm32/stm32mp157a-st-discovery/rtconfig.h [hunk 5]
```diff
@@ -92,6 +110,7 @@
 /* POSIX layer and C standard library */
 
 #define RT_USING_LIBC
+#define RT_USING_POSIX
```
Comment (SummerLife) [inline bsp/stm32/stm32mp157a-st-discovery/rtconfig.h:113]: ~

File: bsp/stm32/libraries/HAL_Drivers/drv_dac.c [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+static struct stm32_dac stm32_dac_obj[sizeof(dac_config) / sizeof(dac_config[0])];
+
+static rt_err_t stm32_dac_enabled(struct rt_dac_device *device, rt_uint32_t channel)
+{
+    DAC_HandleTypeDef *stm32_dac_handler;
+    RT_ASSERT(device != RT_NULL);
+    stm32_dac_handler = device->parent.user_data;
+
+#if defined(SOC_SERIES_STM32MP1)
+        HAL_DAC_Start(stm32_dac_handler, channel);
```
Comment (SummerLife) [inline bsp/stm32/libraries/HAL_Drivers/drv_dac.c:42]: please format the code

File: bsp/stm32/libraries/HAL_Drivers/drv_dac.c [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+static rt_err_t stm32_dac_disabled(struct rt_dac_device *device, rt_uint32_t channel)
+{
+    DAC_HandleTypeDef *stm32_dac_handler;
+    RT_ASSERT(device != RT_NULL);
+    stm32_dac_handler = device->parent.user_data;
+    
+#if defined(SOC_SERIES_STM32MP1)
+        HAL_DAC_Stop(stm32_dac_handler, channel);
```
Comment (SummerLife) [inline bsp/stm32/libraries/HAL_Drivers/drv_dac.c:55]: please format the code

File: bsp/stm32/stm32mp157a-st-discovery/.config (hunk 3 above)
Comment (SummerLife) [inline bsp/stm32/stm32mp157a-st-discovery/.config:111]: It seems that DFS is still enabled.

File: bsp/stm32/stm32mp157a-st-discovery/.config [hunk 8]
```diff
@@ -152,6 +180,10 @@ CONFIG_RT_USING_PIN=y
 #
 CONFIG_RT_USING_LIBC=y
 # CONFIG_RT_USING_PTHREADS is not set
+CONFIG_RT_USING_POSIX=y
```
Comment (SummerLife) [inline bsp/stm32/stm32mp157a-st-discovery/.config:183]: And posix option.

File: components/drivers/include/drivers/dac.h [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+{
+    struct rt_device parent;
+    const struct rt_dac_ops *ops;
+};
+typedef struct rt_dac_device *rt_dac_device_t;
+
+typedef enum
+{
+    RT_dac_CMD_ENABLE,
+    RT_dac_CMD_DISABLE,
```
Comment (whj4674672) [inline components/drivers/include/drivers/dac.h:33]: dac=》DAC

File: components/drivers/misc/dac.c [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+        rt_kprintf("dac write <channel> <value>    - write dac value on the channel\n");
+        rt_kprintf("dac disable <channel>          - disable dac channel\n");
+        rt_kprintf("dac enable <channel>           - enable dac channel\n");
+        result = -RT_ERROR;
+    }
+    return RT_EOK;
+}
+MSH_CMD_EXPORT(dac, dac function);
+
+#endif /* FINSH_USING_MSH */
```
[the remark this answers is not in the document]
Comment (thread-liu) [reply components/drivers/misc/dac.c:231]: 这里是有空格的

File: bsp/stm32/libraries/HAL_Drivers/drv_dac.c [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+                result = -RT_ERROR;
+            }
+        }
+    }
+
+    return result;
+}
+INIT_DEVICE_EXPORT(stm32_dac_init);
+
+#endif /* BSP_USING_DAC */
```
[the remark this answers is not in the document]
Comment (thread-liu) [reply bsp/stm32/libraries/HAL_Drivers/drv_dac.c:188]: 这里是有空格的

**Pull request merged**